test/core/end2end/end2end_test_fuzzer.cc

```diff
@@ -0,0 +1,218 @@
+    // Deduce begin of relevant window.
+    size_t j_begin = 1;
+    if (i > lower_diag) {
+      j_begin = i - lower_diag;
+      d[i][j_begin - 1] = cutoff_plus_1;
+    } else {
+      d[i][0] = static_cast<uint8_t>(i);
+    }
+
+    // Deduce end of relevant window.
+    size_t j_end = i + upper_diag;
+    if (j_end > s2.size()) {
+      j_end = s2.size();
+    } else {
+      d[i][j_end + 1] = cutoff_plus_1;
+    }
+
+    for (size_t j = j_begin; j <= j_end; ++j) {
+      const uint8_t deletion_distance = d[i - 1][j] + 1;
+      const uint8_t insertion_distance = d[i][j - 1] + 1;
+      const uint8_t mismatched_tail_cost = s1[i - 1] == s2[j - 1] ? 0 : 1;
+      const uint8_t mismatch_distance = d[i - 1][j - 1] + mismatched_tail_cost;
+      uint8_t transposition_distance = _cutoff + 1;
+      if (i > 1 && j > 1 && s1[i - 1] == s2[j - 2] && s1[i - 2] == s2[j - 1])
+        transposition_distance = d[i - 2][j - 2] + 1;
+      d[i][j] = std::min({cutoff_plus_1, deletion_distance, insertion_distance,
+                          mismatch_distance, transposition_distance});
+    }
+  }
+  return d[s1.size()][s2.size()];
+}
+}  // namespace
+
+DEFINE_PROTO_FUZZER(const core_end2end_test_fuzzer::Msg& msg) {
+  grpc_core::g_is_fuzzing_core_e2e_tests = true;
+
+  struct Test {
+    std::string name;
+    absl::AnyInvocable<std::unique_ptr<grpc_core::CoreEnd2endTest>() const>
+        factory;
+  };
+
+  static const auto all_tests =
+      grpc_core::CoreEnd2endTestRegistry::Get().AllTests();
+  static const auto tests = []() {
+    auto only_suite = grpc_core::GetEnv("GRPC_TEST_FUZZER_SUITE");
+    auto only_test = grpc_core::GetEnv("GRPC_TEST_FUZZER_TEST");
+    auto only_config = grpc_core::GetEnv("GRPC_TEST_FUZZER_CONFIG");
+    std::vector<Test> tests;
+    for (const auto& test : all_tests) {
+      if (test.config->feature_mask & FEATURE_MASK_DO_NOT_FUZZ) continue;
+      if (only_suite.has_value() && test.suite != only_suite.value()) continue;
+      if (only_test.has_value() && test.name != only_test.value()) continue;
+      if (only_config.has_value() && test.config->name != only_config.value()) {
+        continue;
+      }
+      tests.emplace_back(
+          Test{absl::StrCat(test.suite, ".", test.name, "/", test.config->name),
+               [&test]() {
+                 return std::unique_ptr<grpc_core::CoreEnd2endTest>(
+                     test.make_test(test.config));
+               }});
+    }
+    GPR_ASSERT(tests.size() > 0);
+    return tests;
+  }();
+  static const auto only_experiment =
+      grpc_core::GetEnv("GRPC_TEST_FUZZER_EXPERIMENT");
+
+  if (squelch && !grpc_core::GetEnv("GRPC_TRACE_FUZZER").has_value()) {
+    gpr_set_log_function(dont_log);
+  }
+
+  auto test_name =
+      absl::StrCat(msg.suite(), ".", msg.test(), "/", msg.config());
+  size_t best_test = 0;
+  for (size_t i = 0; i < tests.size(); i++) {
+    if (CappedDamerauLevenshteinDistance(test_name, tests[i].name, 100) <
```

Comment:
Not sure if the fuzzer will test every test config with equal probability. Perhaps in a future PR, we might want to explore a simpler scheme (e.g fuzzer generating random indexes, instead of strings etc)

Reply:
yeah I'm super torn on this, and have been through several iterations.
It's *very* useful to be able to lock a specific test, and the edit distance approach helps here.
I've been careful to include a corpus that contains all test names for the fuzzer to start with, so hopefully that helps.